### PR TITLE
Normalize bound-identifier stamps when saving a .cmi

### PR DESCRIPTION
**Pull request by riaqn**
The signature in a `cmi` contains `Ident.t` that are currently based on a global stamp counter, which is how many identifiers happened to be allocated while compiling it — which is perturbed by things such as loading its dependencies. This makes the byte representation of the signature volatile even if the signature "morally" stays the same. This defeats content-based build cutoff and cascading spurious recompilations.

This PR makes the signature in a `cmi` less volatile, by restamping the `Ident.t` when the signature is saved.

- [x] check performance degeneration

### CODE REVIEW

**Comment by riaqn** [comment]
Benchmarking on internal code base (990 files, 10 iterations) shows a ~1% slowdown.

EDIT: I did another benchmark on the latest commit and it comes out ~0.15% slowdown. I don't think any of our code changes substantially, so the numbers are probably mostly noise anyway.

File: typing/subst.ml [hunk 1]
```diff
@@ -307,7 +311,13 @@ let with_additional_action =
         let prepare_modality modality =
           Mode.Modality.(modality |> to_const_exn|> of_const)
         in
-        Prepare_for_saving { prepare_jkind; prepare_mode; prepare_modality },
+        let bound_ident_stamp = ref 0 in
```
Comment (ainageorges) [inline typing/subst.ml:311]: We here create a local `stamp` counter, and define an alternate `rename` via `prepare_ident` that encloses it. This looks correct, and it indeed means the stamp counter is reset whenever `with_additional_action Prepare_for_saving` is called. It is however quite different from how the `id` counter is reset, which maintains a global id counter, and resets it in `reset_additional_action_id`. Could there be an advantage in applying a similar design? (using the same stamp counter, and reset it in `reset_additional_action_id`)

File: typing/subst.ml
Comment (ainageorges) [inline typing/subst.ml]: I believe this is no longer true: the `prepare_ident` function in `Prepare_for_saving` encloses different counters, although they should be observationally equivalent.

File: typing/subst.ml (hunk 1 above)
Comment (riaqn) [reply typing/subst.ml:311]: I think it might be better to adopt a uniform design throughout. However, in that case I'm inclined to think that we should convert the `id` counter to our design instead of the other way around, because it's usually better to avoid global states. For example, if the compiler is outputting multiple `cmi` files, you might forget to reset the global counter, while our design forces you to create a new counter.

Comment (ainageorges) [reply typing/subst.ml:311]: That's fair yes. Maybe we should define a record that contains all the various counters, and make `reset_additional_action_id` a constructor rather than a reset function? Alternatively we can keep the global counters and call `reset` from `with_additional_action`

File: testsuite/tests/cmi-stamp-normalization/test.ml [hunk 2]
```diff
@@ -42,5 +42,5 @@ bound_ident_stamps "ocamlc.byte/bar.cmi"
 ;;
 [%%expect {|
 val bound_ident_stamps : string -> string list = <fun>
-- : string list = ["t_288"; "y_289"]
+- : string list = ["t_1"; "y_2"]
```
Comment (ainageorges) [inline testsuite/tests/cmi-stamp-normalization/test.ml]: Can we elaborate on this test to show the volatile behavior with two different loading dependencies of the same signature?

Comment (ainageorges) [reply testsuite/tests/cmi-stamp-normalization/test.ml]: For example, I think if we define a `bar_v2.mli` which is byte identical to `bar.mli`, and call `bound_ident_stamps "ocamlc.byte/bar_v2.cmi"` on it, we will get different stamps on the output, but after the fix they will have the same stamps.

File: testsuite/tests/cmi-stamp-normalization/bar.mli [hunk 3]
```diff
@@ -0,0 +1,3 @@
+type t = Foo.t
+
```
Comment (ainageorges) [inline testsuite/tests/cmi-stamp-normalization/bar.mli:2]: We should add a record type and variant type to assert that they remain distinct

File: typing/subst.ml (hunk 1 above)
Comment (riaqn) [reply typing/subst.ml:311]: so it turns out that type id needs to be updated not only for `Prepare_for_saving` but also for `Duplicate_variables`. Therefore it's not suitable to move this logic into `Prepare_for_saving`. 

Also `reset_additional_actions_id` and `with_additional_action` are not always called together.

I suggest we just keep as it is.

Comment (ainageorges) [reply typing/subst.ml:311]: Ah yes, `reset_additional_action_id` is always called, whereas `with_additional_action Prepare_for_saving` is called lazily. Let's keep it is as is then

Comment (riaqn) [reply typing/subst.ml:311]: So it seems that we can separate `Prepare_for_saving` and `Duplicate_variables` and make the latter still use a global id counter. I tried and it works, but involves lots of changes in subst.ml that I think might not be worthwhile.

Comment (ainageorges) [reply typing/subst.ml:311]: Indeed I don't think that is worthwhile. For what it's worth, I think the global counters are fine for the id's, they are well documented and resetting them is done only once.

File: testsuite/tests/cmi-stamp-normalization/test.ml [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+let bound_ident_stamps cmi_file =
+  let cmi = Cmi_format.read_cmi cmi_file in
+  let sg, _ = cmi.Cmi_format.cmi_sign in
+  List.map
+    (fun (item : Types.signature_item) ->
+      let id =
+        match item with
+        | Sig_value (id, _, _) -> id
+        | Sig_type (id, _, _, _) -> id
```
Comment (ainageorges) [inline testsuite/tests/cmi-stamp-normalization/test.ml]: Here we can also print the label declaration defined within the second projection of `Sig_type`, to assert that the new counter does not affect record labels

**Review by ainageorges — APPROVED** [review]
I only have one final suggestion, otherwise this looks good to me!

File: testsuite/tests/cmi-stamp-normalization/test.ml (hunk 4 above)
Comment (ainageorges) [reply testsuite/tests/cmi-stamp-normalization/test.ml]: As a follow up: should these in fact be unique stamps, or should they be the same stamp? It's unclear to me if it's sound for them to have the same stamp, but perhaps the stamp is not used at all for record fields, in which case we probably want them to be same to prevent recompilation.

**Review by ainageorges — CHANGES_REQUESTED** [review]
We should make sure that either 
1) stamp collisions are no issue (between label `Ident.t` and others) or 
2) do we even need `Ident.t` here? by looking at it offline together, it appears as if only the string part is used. As discussed offline, we should confirm whether we can replace them with strings, which would simplify the code.

File: testsuite/tests/cmi-stamp-normalization/test.ml (hunk 4 above)
Comment (riaqn) [reply testsuite/tests/cmi-stamp-normalization/test.ml]: the field names and constructor names are printed now.

`cd_id`'s stamp are used in corner cases, which I'd rather not change for simplicity. The latest commit just restamp all of them to make the stamps more stable.

**Review by ainageorges — APPROVED** [review]
The new code looks good, and the result shows stable record fields and variant constructors. It's a shame we could not remove the stamp entirely, but this solution look good as well.

File: typing/subst.ml [hunk 5]
```diff
@@ -796,9 +803,23 @@ let type_expr s ty =
...
+  match s.additional_action with
+  | Prepare_for_saving { prepare_ident; _ } -> prepare_ident id
+  | Duplicate_variables | No_action -> Ident.rename id
+
+(* For constructor/label idents, which may be predef (e.g. the constructors
+   of [bool] or [or_null], whose declarations can appear in a substituted
+   signature). Predef and global idents are kept: they cannot be renamed,
+   and they carry no volatile stamp. *)
+let rename_decl_ident s id =
+  if Ident.is_global_or_predef id then id else rename_ident s id
```
Comment (ainageorges) [inline typing/subst.ml:818]: Since some identifiers get renamed and others are not, it means we could be unlucky and get the same stamp. Could this be an issue?

**Review by ainageorges — CHANGES_REQUESTED** [review]

File: typing/subst.ml (hunk 5 above)
Comment (riaqn) [reply typing/subst.ml:818]: As discussed offline, the stamps in `Predef` is distinct from the stamps in `Local` (different counter) and thus this PR doesn't change the behavior wrt `Predef` in anyway.

Comment (ainageorges) [reply typing/subst.ml:818]: Agreed, thank you for clarifying!

**Review by ainageorges — APPROVED** [review]
